File: DSA/src/Questions/BinarySearch/KthSmallestElementInMatrix.java
package Questions.BinarySearch;

/*
https://leetcode.com/problems/kth-smallest-element-in-a-sorted-matrix/description/
 Given an n x n matrix where each of the rows and columns is sorted in ascending order, return the kth smallest element in the matrix.

Note that it is the kth smallest element in the sorted order, not the kth distinct element.

You must find a solution with a memory complexity better than O(n2).

 

Example 1:

Input: matrix = [[1,5,9],[10,11,13],[12,13,15]], k = 8
Output: 13
Explanation: The elements in the matrix are [1,5,9,10,11,12,13,13,15], and the 8th smallest number is 13
Example 2:

Input: matrix = [[-5]], k = 1
Output: -5
 

Constraints:

n == matrix.length == matrix[i].length
1 <= n <= 300
-109 <= matrix[i][j] <= 109
All the rows and columns of matrix are guaranteed to be sorted in non-decreasing order.
1 <= k <= n2
 Time Complexity: O(n * log(max - min))
    Space Complexity: O(1)


    low = matrix[0][0]
    high = matrix[n - 1][n - 1]
    while low < high:
        mid = (low + high) // 2
        count = countLessEqual(matrix, mid, n)
        if count < k:
            low = mid + 1
        else:
            high = mid
    return low

    countLessEqual(matrix, target, n):
        count = 0
        row = n - 1
        col = 0
        while row >= 0 and col < n:
            if matrix[row][col] <= target:
                count += row + 1
                col += 1
            else:
                row -= 1
        return count

 */
public class KthSmallestElementInMatrix {
    public int kthSmallest(int[][] matrix, int k) {
        int n = matrix.length;
        int low = matrix[0][0];
        int high = matrix[n - 1][n - 1];

        while (low < high) {
            int mid = (high + low) >> 1;
            int count = countLessEqual(matrix, mid, n);

            if (count < k) {
                low = mid + 1;
            } else {
                high = mid;
            }
        }

        return low;
    }

    private int countLessEqual(int[][] matrix, int target, int n) {
        int count = 0;
        int row = n - 1;
        int col = 0;

        while (row >= 0 && col < n) {
            if (matrix[row][col] <= target) {
                count += (row + 1);
                col++;
            } else {
                row--;
            }
        }

        return count;
    }

    public static void main(String[] args) {
        KthSmallestElementInMatrix obj = new KthSmallestElementInMatrix();
        int[][] matrix = {
                { 1, 5, 9 },
                { 10, 11, 13 },
                { 12, 13, 15 }
        };
        int k = 8;
        System.out.println(obj.kthSmallest(matrix, k)); // Output: 13
    }

}

File: DSA/Python/Questions/BinarySearch/KthSmallestElementInMatrix.py
"""
# Question: Kth Smallest Element in a Sorted Matrix
# Link: https://leetcode.com/problems/kth-smallest-element-in-a-sorted-matrix/

# Problem Statement:
# Given an n x n matrix where each row and column is sorted in ascending order,
# find the kth smallest element in the matrix.

# Example:
# Input: matrix = [[1,5,9],[10,11,13],[12,13,15]], k = 8
# Output: 13
# Explanation: The elements in sorted order are [1,5,9,10,11,12,13,13,15]
"""

from typing import List
import heapq


class KthSmallestElementInMatrix:
    def kth_smallest_binary(self, matrix: List[List[int]], k: int) -> int:
        """
        Binary search approach using count of elements less than mid
        Time Complexity: O(n * log(max-min))
        Space Complexity: O(1)
        """
        n = len(matrix)
        left, right = matrix[0][0], matrix[n - 1][n - 1]

        while left < right:
            mid = (left + right) >> 1
            count = self.count_less_equal(matrix, mid)

            if count < k:
                left = mid + 1
            else:
                right = mid

        return left

    def count_less_equal(self, matrix: List[List[int]], target: int) -> int:
        """Helper function to count elements less than or equal to target"""
        count = 0
        n = len(matrix)
        col = n - 1

        for row in range(n):
            while col >= 0 and matrix[row][col] > target:
                col -= 1
            count += col + 1

        return count

    def kth_smallest_heap(self, matrix: List[List[int]], k: int) -> int:
        """
        Min-heap approach
        Time Complexity: O(k * log(n))
        Space Complexity: O(n)
        """
        n = len(matrix)
        heap = [(matrix[0][0], 0, 0)]
        visited = {(0, 0)}

        for _ in range(k - 1):
            val, row, col = heapq.heappop(heap)

            for new_row, new_col in [(row + 1, col), (row, col + 1)]:
                if new_row < n and new_col < n and (new_row, new_col) not in visited:
                    heapq.heappush(heap, (matrix[new_row][new_col], new_row, new_col))
                    visited.add((new_row, new_col))

        return heapq.heappop(heap)[0]

    def kth_smallest_flatten(self, matrix: List[List[int]], k: int) -> int:
        """
        Flatten and sort approach (less efficient but simple)
        Time Complexity: O(n^2 * log(n^2))
        Space Complexity: O(n^2)
        """
        flattened = []
        for row in matrix:
            flattened.extend(row)
        return sorted(flattened)[k - 1]


def main():
    test_cases = [
        {"matrix": [[1, 5, 9], [10, 11, 13], [12, 13, 15]], "k": 8, "expected": 13},
        {"matrix": [[1, 2], [1, 3]], "k": 2, "expected": 1},
        {"matrix": [[-5]], "k": 1, "expected": -5},
    ]

    solution = KthSmallestElementInMatrix()
    for i, test in enumerate(test_cases):
        print(f"\nTest Case {i + 1}:")
        print(f"Matrix: {test['matrix']}")
        print(f"k: {test['k']}")

        result_binary = solution.kth_smallest_binary(test["matrix"], test["k"])
        result_heap = solution.kth_smallest_heap(test["matrix"], test["k"])
        result_flatten = solution.kth_smallest_flatten(test["matrix"], test["k"])

        print(f"Binary Search Result: {result_binary}")
        print(f"Heap Result: {result_heap}")
        print(f"Flatten Result: {result_flatten}")
        print(f"Expected: {test['expected']}")

        assert result_binary == result_heap == result_flatten == test["expected"]


if __name__ == "__main__":
    main()
